coax/td_learning/_clippeddoubleqlearning.py

```diff
@@ -139,143 +138,101 @@ def __init__(
-
-            # target-network estimate (is this worth computing?)
-            Q_targ_list = []
-            qs = list(zip(self.q_targ_list, target_params['q_targ'], target_state['q_targ']))
-            for q, pm, st in qs:
-                Q_targ, _ = q.function_type1(pm, st, next(rngs), S, A, False)
-                assert Q_targ.ndim == 1, f"bad shape: {Q_targ.shape}"
-                Q_targ_list.append(Q_targ)
-            Q_targ_list = jnp.stack(Q_targ_list, axis=-1)
-            assert Q_targ_list.ndim == 2, f"bad shape: {Q_targ_list.shape}"
-            Q_targ = jnp.min(Q_targ_list, axis=-1)
-
-            chex.assert_equal_shape([td_error, W, Q_targ])
-            metrics = {
-                f'{self.__class__.__name__}/loss': loss,
-                f'{self.__class__.__name__}/td_error': jnp.mean(W * td_error),
-                f'{self.__class__.__name__}/td_error_targ': jnp.mean(-dLoss_dQ(Q, Q_targ, W)),
-            }
-            return loss, (td_error, state_new, metrics)
+                Q_sa_targets = targets
+            assert Q_sa_targets.ndim == 2
+            return jnp.argmin(Q_sa_targets, axis=0)
 
         def grads_and_metrics_func(
                 params, target_params, state, target_state, rng, transition_batch):
-
             rngs = hk.PRNGSequence(rng)
-            grads, (td_error, state_new, metrics) = jax.grad(loss_func, has_aux=True)(
-                params, target_params, state, target_state, next(rngs), transition_batch)
-
-            # add some diagnostics about the gradients
-            metrics.update(get_grads_diagnostics(grads, f'{self.__class__.__name__}/grads_'))
-
-            return grads, state_new, metrics, td_error
+            # compute argmin target q-function per transition
+            q_targ_argmin = q_targ_pi_argmin(
+                target_params, target_state, next(rngs), transition_batch)
+
+            def apply_params(q_targ_idx, transition, p, s):
+                # extract target q-function parameters and state
+                t_params = jax.tree_util.tree_map(lambda t: t[q_targ_idx], p)
+                t_state = jax.tree_util.tree_map(lambda t: t[q_targ_idx], s)
+                return self._super_grads_and_metrics_func(params, t_params,
+                                                          state, t_state,
+                                                          next(rngs), single_to_batch(transition))
+            single_grads_and_metrics = jax.vmap(
+                apply_params, in_axes=(0, 0, None, None), out_axes=0)
+            # have to average because the grads are computed using a single transition
+            batch_grads, batch_state_new, batch_metrics, batch_td_error = jax.tree_util.tree_map(
+                lambda t: jnp.mean(t, axis=0), single_grads_and_metrics(
+                    q_targ_argmin, transition_batch, *stack_trees(target_params, target_state))
+            )
+
+            return batch_grads, batch_state_new, batch_metrics, batch_td_error
```

Comment:
I will have to look at this later.. it's late and I'm a bit too tired to wrap my head around it.

Reply:
Yes, this part was new territory for me. What I think it does (should do?) is that the argmin `q_targ` indices and corresponding transitions are vmapped together with the `q_targ` parameters and states. 

However, this way the metrics, gradients and TD errors have to be averaged because they were computed on a batch size of 1. I think averaging the state might be wrong, now that I look at it. At least for state that is not averaged over a batch (e.g. summed or something else). But I also do not see a better way, how to solve this, yet.